File: packages/contracts/src/narrative.ts
import type { GameId, IsoDateTime, JsonObject, Uuid } from "./common.js";

export type NarrativeBeatStatus = "found" | "partially_found" | "ambiguous" | "missing" | "contradicted" | "unusable";
export type ClaimStatus = "hypothesis" | "observed_once" | "reproduced" | "verified" | "contradicted" | "outdated" | "unknown";
export type EvidenceSourceType = "segment" | "media_frame" | "ocr_text" | "detected_entity" | "detected_event" | "knowledge_item" | "official_documentation" | "repeated_test" | "user_library";

export interface SegmentCandidate {
  readonly segment_id: Uuid;
  readonly start_ms: number;
  readonly end_ms: number;
  readonly score: number;
  readonly rationale: string;
}

export interface NarrativeBeat {
  readonly id: Uuid;
  readonly order: number;
  readonly intent: string;
  readonly required: boolean;
  readonly status: NarrativeBeatStatus;
  readonly candidate_segments: readonly SegmentCandidate[];
  readonly concept?: string;
  readonly purpose?: ScriptPurpose;
  readonly explicitly_requested?: boolean;
  readonly decision_reason?: string;
}

export interface NarrativeMap {
  readonly id: Uuid;
  readonly project_id: Uuid;
  readonly brief_id: Uuid;
  readonly version: number;
  readonly beats: readonly NarrativeBeat[];
  readonly required_coverage: number;
  readonly missing_required_count: number;
  readonly overall_coverage?: number;
  readonly content_type?: string;
  readonly algorithm_version?: string;
  readonly fact_boundary?: string;
  readonly created_at: IsoDateTime;
}

export interface MissingSequenceRecommendation {
  readonly beat_id: Uuid;
  readonly intent: string;
  readonly priority: "required" | "recommended";
  readonly request: string;
  readonly content_type: string;
  readonly proof_goal: string;
}

export interface RequestedFactGate {
  readonly request: string;
  readonly status: "requires_phase5_verification";
  readonly allowed_in_script: false;
}

export interface CoverageReport {
  readonly id: Uuid;
  readonly narrative_map_id: Uuid;
  readonly project_id: Uuid;
  readonly brief_id: Uuid;
  readonly required_coverage: number;
  readonly overall_coverage: number;
  readonly mandatory_total: number;
  readonly mandatory_found: number;
  readonly missing_items: readonly Record<string, unknown>[];
  readonly ambiguous_items: readonly Record<string, unknown>[];
  readonly low_quality_sequences: readonly Record<string, unknown>[];
  readonly requested_facts: readonly RequestedFactGate[];
  readonly complementary_footage: readonly MissingSequenceRecommendation[];
  readonly editing_decision: "ready_with_prudent_narration" | "continue_adapted_with_warning" | "continue_partial_and_request_footage";
  readonly created_at: IsoDateTime;
}

export type ContentPlanVariant = "direct" | "storytelling" | "very_dynamic";

export interface ContentPlan {
  readonly id: Uuid;
  readonly variant: ContentPlanVariant;
  readonly selected: boolean;
  readonly score: number;
  readonly description: string;
  readonly selection_reason: string;
  readonly beats: readonly Record<string, unknown>[];
  readonly selection_signals: Readonly<Record<string, number | string>>;
}

export interface Claim {
  readonly id: Uuid;
  readonly project_id: Uuid;
  readonly game_id: GameId;
  readonly claim_key: string;
  readonly claim_type: string;
  readonly statement: string;
  readonly normalized_statement: string;
  readonly status: ClaimStatus;
  readonly confidence: number;
  readonly game_version: string | null;
  readonly observed_at: IsoDateTime | null;
  readonly verified_at: IsoDateTime | null;
  readonly allowed_in_script: boolean;
  readonly certainty_language: string;
  readonly verification_reason: string;
  readonly safe_narration: string | null;
  readonly request: string | null;
  readonly algorithm_version: string;
  readonly evidence: readonly EvidenceRecord[];
}

export interface EvidenceRecord {
  readonly id: Uuid;
  readonly evidence_type: EvidenceSourceType;
  readonly source_id: Uuid;
  readonly start_ms: number | null;
  readonly end_ms: number | null;
  readonly strength: number;
  readonly metadata: JsonObject;
}

export interface RequestedFactResult {
  readonly claim_id: Uuid;
  readonly request: string;
  readonly status: ClaimStatus;
  readonly confidence: number;
  readonly allowed_in_script: boolean;
  readonly reason: string;
}

export interface VerificationReport {
  readonly id: Uuid;
  readonly schema_version: "1.0";
  readonly algorithm_version: string;
  readonly project_id: Uuid;
  readonly brief_id: Uuid;
  readonly game_id: GameId;
  readonly status: "PASSED" | "PASSED_WITH_EXCLUSIONS" | "FAILED";
  readonly claims: readonly Claim[];
  readonly summary: {
    readonly claim_count: number;
    readonly admitted_claim_count: number;
    readonly blocked_claim_count: number;
    readonly requested_fact_count: number;
    readonly requested_fact_coverage: number;
    readonly script_factual_safety: number;
    readonly status_distribution: Readonly<Record<ClaimStatus, number>>;
    readonly knowledge_items_available: number;
    readonly knowledge_items_used: number;
  };
  readonly requested_facts: readonly RequestedFactResult[];
  readonly knowledge_snapshot: {
    readonly namespace: GameId;
    readonly game_version: string;
    readonly item_count: number;
    readonly used_item_ids: readonly Uuid[];
    readonly cross_game_items: number;
    readonly notice: string;
  };
  readonly gate: {
    readonly status: "PASSED" | "PASSED_WITH_EXCLUSIONS" | "FAILED";
    readonly rule: string;
    readonly blocked_claim_ids: readonly Uuid[];
    readonly admitted_claim_ids: readonly Uuid[];
  };
  readonly created_at: IsoDateTime;
}

export interface KnowledgeItem {
  readonly id: Uuid;
  readonly namespace: "gta5" | "gta6";
  readonly canonical_key: string;
  readonly game_id: "gta5" | "gta6";
  readonly game_version: string;
  readonly value: JsonObject;
  readonly source_uri: string | null;
  readonly source_type: string;
  readonly confidence: number;
  readonly status: ClaimStatus;
  readonly verified_at: IsoDateTime | null;
  readonly revision: number;
}

export type ScriptPurpose = "hook" | "context" | "explanation" | "transition" | "proof" | "comparison" | "conclusion" | "call_to_action";

export interface ScriptBlock {
  readonly id: Uuid;
  readonly order: number;
  readonly purpose: ScriptPurpose;
  readonly narration: string;
  readonly on_screen_text: string | null;
  readonly supporting_segment_ids: readonly Uuid[];
  readonly supporting_claim_ids: readonly Uuid[];
  readonly estimated_duration_ms: number;
  readonly confidence: number;
}

File: legacy/packages/contracts/python/src/gta_ai_contracts/narrative.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field, model_validator

from .common import GameId, JsonValue, NonNegativeInt, PositiveInt, StudioModel, UnitScore


class SegmentCandidate(StudioModel):
    segment_id: UUID
    start_ms: NonNegativeInt
    end_ms: PositiveInt
    score: UnitScore
    rationale: str

    @model_validator(mode="after")
    def validate_range(self) -> "SegmentCandidate":
        if self.end_ms <= self.start_ms:
            raise ValueError("candidate end must be greater than start")
        return self


class NarrativeBeat(StudioModel):
    id: UUID
    order: NonNegativeInt
    intent: str
    required: bool
    status: Literal["found", "partially_found", "ambiguous", "missing", "contradicted", "unusable"]
    candidate_segments: list[SegmentCandidate] = Field(default_factory=list)
    concept: str | None = None
    purpose: Literal["hook", "context", "explanation", "transition", "proof", "comparison", "conclusion", "call_to_action"] | None = None
    explicitly_requested: bool = False
    decision_reason: str | None = None

    @model_validator(mode="after")
    def validate_candidates(self) -> "NarrativeBeat":
        if self.status == "found" and not self.candidate_segments:
            raise ValueError("found beats require at least one candidate")
        if self.status == "missing" and self.candidate_segments:
            raise ValueError("missing beats cannot contain candidates")
        return self


class NarrativeMap(StudioModel):
    id: UUID
    project_id: UUID
    brief_id: UUID
    version: PositiveInt
    beats: list[NarrativeBeat]
    required_coverage: UnitScore
    missing_required_count: NonNegativeInt
    overall_coverage: UnitScore | None = None
    content_type: str | None = None
    algorithm_version: str | None = None
    fact_boundary: str | None = None
    created_at: datetime


class MissingSequenceRecommendation(StudioModel):
    beat_id: UUID
    intent: str
    priority: Literal["required", "recommended"]
    request: str
    content_type: str
    proof_goal: str


class RequestedFactGate(StudioModel):
    request: str
    status: Literal["requires_phase5_verification"]
    allowed_in_script: Literal[False]


class CoverageReport(StudioModel):
    id: UUID
    narrative_map_id: UUID
    project_id: UUID
    brief_id: UUID
    required_coverage: UnitScore
    overall_coverage: UnitScore
    mandatory_total: NonNegativeInt
    mandatory_found: NonNegativeInt
    missing_items: list[dict[str, object]] = Field(default_factory=list)
    ambiguous_items: list[dict[str, object]] = Field(default_factory=list)
    low_quality_sequences: list[dict[str, object]] = Field(default_factory=list)
    requested_facts: list[RequestedFactGate] = Field(default_factory=list)
    complementary_footage: list[MissingSequenceRecommendation] = Field(default_factory=list)
    editing_decision: Literal["ready_with_prudent_narration", "continue_adapted_with_warning", "continue_partial_and_request_footage"]
    created_at: datetime


class ContentPlan(StudioModel):
    id: UUID
    variant: Literal["direct", "storytelling", "very_dynamic"]
    selected: bool
    score: UnitScore
    description: str
    selection_reason: str
    beats: list[dict[str, object]] = Field(default_factory=list)
    selection_signals: dict[str, float | str] = Field(default_factory=dict)


class EvidenceRef(StudioModel):
    id: UUID
    evidence_type: Literal[
        "segment", "media_frame", "ocr_text", "detected_entity", "detected_event",
        "knowledge_item", "official_documentation", "repeated_test", "user_library",
    ]
    source_id: UUID
    start_ms: NonNegativeInt | None = None
    end_ms: PositiveInt | None = None
    strength: UnitScore
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class Claim(StudioModel):
    id: UUID
    project_id: UUID
    game_id: GameId
    claim_key: str
    claim_type: str
    statement: str
    normalized_statement: str
    status: Literal["hypothesis", "observed_once", "reproduced", "verified", "contradicted", "outdated", "unknown"]
    confidence: UnitScore
    game_version: str | None = None
    observed_at: datetime | None = None
    verified_at: datetime | None = None
    allowed_in_script: bool
    certainty_language: str
    verification_reason: str
    safe_narration: str | None = None
    request: str | None = None
    algorithm_version: str
    evidence: list[EvidenceRef] = Field(default_factory=list)


class RequestedFactResult(StudioModel):
    claim_id: UUID
    request: str
    status: Literal["hypothesis", "observed_once", "reproduced", "verified", "contradicted", "outdated", "unknown"]
    confidence: UnitScore
    allowed_in_script: bool
    reason: str


class VerificationSummary(StudioModel):
    claim_count: NonNegativeInt
    admitted_claim_count: NonNegativeInt
    blocked_claim_count: NonNegativeInt
    requested_fact_count: NonNegativeInt
    requested_fact_coverage: UnitScore
    script_factual_safety: UnitScore
    status_distribution: dict[str, NonNegativeInt]
    knowledge_items_available: NonNegativeInt
    knowledge_items_used: NonNegativeInt


class KnowledgeSnapshot(StudioModel):
    namespace: GameId
    game_version: str
    item_count: NonNegativeInt
    used_item_ids: list[UUID] = Field(default_factory=list)
    cross_game_items: NonNegativeInt
    notice: str


class FactGate(StudioModel):
    status: Literal["PASSED", "PASSED_WITH_EXCLUSIONS", "FAILED"]
    rule: str
    blocked_claim_ids: list[UUID] = Field(default_factory=list)
    admitted_claim_ids: list[UUID] = Field(default_factory=list)


class VerificationReport(StudioModel):
    id: UUID
    schema_version: Literal["1.0"]
    algorithm_version: str
    project_id: UUID
    brief_id: UUID
    game_id: GameId
    status: Literal["PASSED", "PASSED_WITH_EXCLUSIONS", "FAILED"]
    claims: list[Claim] = Field(default_factory=list)
    summary: VerificationSummary
    requested_facts: list[RequestedFactResult] = Field(default_factory=list)
    knowledge_snapshot: KnowledgeSnapshot
    gate: FactGate
    created_at: datetime


class KnowledgeItem(StudioModel):
    id: UUID
    namespace: Literal["gta5", "gta6"]
    canonical_key: str
    game_id: Literal["gta5", "gta6"]
    game_version: str
    value: dict[str, JsonValue]
    source_uri: str | None = None
    source_type: str
    confidence: UnitScore
    status: Literal["hypothesis", "observed_once", "reproduced", "verified", "contradicted", "outdated", "unknown"]
    verified_at: datetime | None = None
    revision: PositiveInt


class ScriptBlock(StudioModel):
    id: UUID
    order: NonNegativeInt
    purpose: Literal["hook", "context", "explanation", "transition", "proof", "comparison", "conclusion", "call_to_action"]
    narration: str
    on_screen_text: str | None = None
    supporting_segment_ids: list[UUID] = Field(default_factory=list)
    supporting_claim_ids: list[UUID] = Field(default_factory=list)
    estimated_duration_ms: PositiveInt
    confidence: UnitScore
